### ISSUE TITLE
Clearing value of `ui.editor` does not work on 2nd try

### ISSUE THREAD

**Issue by falkoschindler** [zauberzeug/nicegui#3217]
> I'm currently having the exact same problem with ui.editor. editor.set_value(None) works once and then after that not anymore. Any suggestions for this one?

_Originally posted by @tjongsma in https://github.com/zauberzeug/nicegui/issues/1088#issuecomment-2162768878_

Reproduction:

```py
editor = ui.editor().on('keypress.enter', lambda: editor.set_value(None))
```

**Comment by frankvp11** [comment]
This is quite an interesting problem.
Something deeper is going on than just the editor....
```python
from nicegui import ui


def handle_event1():
    editor.set_value(None)
    print(editor._props)


def handle_event2():
    print(editor._props)


editor = ui.editor().on("keypress.enter", lambda: handle_event1())
ui.button("Clear", on_click=handle_event2)


ui.run()
```

When I run something like this, I can see that the props aren't actually being updated... 
When I type something, then press enter, then press the button, it shows me that the model-value is still set to the text I typed.
This is the reason that when you press enter on the second time it doesn't remove it, or atleast that's what I'm thinking.

I think solving this issue will come down to figuring out why the props aren't actually updated

**Comment by tjongsma** [comment]
Thanks for looking into this!As a temporary fix for anyone coming across this, I sort of solved the issue with a bit of a bandaid. My intuition was that the issue was that it only worked once because somehow setting the value would only work if it was a new value and something wasn't being updated right (which I think you're also finding). So what worked for me was some simple logic:flag=0Def clear_editor():Global flagIf flag==0: Editor.set_value(none) Flag=1Else: Editor.set_value('') Flag=0Sorry for the formatting, I'm on my phone right now. But this seems to work, at least visually.On Jun 20, 2024 12:03, frankvp ***@***.***> wrote:
This is quite an interesting problem.
Something deeper is going on than just the editor....
from nicegui import ui


def handle_event1():
    editor.set_value(None)
    print(editor._props)


def handle_event2():
    print(editor._props)


editor = ui.editor().on("keypress.enter", lambda: handle_event1())
ui.button("Clear", on_click=handle_event2)


ui.run()
When I run something like this, I can see that the props aren't actually being updated...
When I type something, then press enter, then press the button, it shows me that the model-value is still set to the text I typed.
This is the reason that when you press enter on the second time it doesn't remove it, or atleast that's what I'm thinking.
I think solving this issue will come down to figuring out why the props aren't actually updated

**Comment by frankvp11** [comment]
To help us all, I've figured out what @tjongsma wanted (in brief)
```python

from nicegui import ui

@ui.page("/")
def main():
    flag = False
    def handle_event(e):
        nonlocal flag
        if flag:
            editor.set_value("")
            flag = not flag
        else:
            editor.set_value(None)
            flag = not flag
    
        editor.update()

    def print_dict():
        print(editor._to_dict()['props'])


    editor = ui.editor().on('keypress.enter', handle_event)
    ui.button('Print dict', on_click=print_dict)

ui.run()
```


We should look into seeing if there might be a better way to handle the setting of None for the editor. Perhaps we should into some core functionality of the ValueElement and how it handles None. The reason that I say this is because despite the fact that this fix that I gave above does technically work, it doens't solve the underlying issue which is that the props aren't properly being cleared. Despite this fix somehow working, after typing some text, then clearing it from the editor, when I run the print_dict() function by the button press, it still shows the editor as having the old value in the model-value. This is something that should be fixed.

**Comment by frankvp11** [comment]
I think I've figured out the bug, and I don't believe that it's Nicegui's fault.
I would appreciate if someone could verify this somehow, but basically, here is what I think.
TLDR; race condition.
Basically, when you press the enter key, it sends two events. One get's senter to the editor element which thinks that you want to place content on the newline. The other get's picked up by us setting the content to None. The content first get's set to None, then the event get's registered from the element itself and it tries to set the content to the previous, with the newline. 
This causes a race condition in some sense because it has to do with the timing.
This is what happens when I print out the contents of the message [here](https://github.com/zauberzeug/nicegui/blob/e4471e5dd0c558d019545c193b1374002ef69695/nicegui/element.py#L460-L464)
 ```python
{'id': 4, 'client_id': 'a48113b2-a22d-427d-8bbf-601e4e0500fe', 'listener_id': '650d15a8-6afe-43bf-bfc5-72aa7bdb6627', 'args': 'a'}
{'id': 4, 'client_id': 'a48113b2-a22d-427d-8bbf-601e4e0500fe', 'listener_id': '72ed4040-9518-4a02-b7f1-b13400fca0a5', 'args': {'isTrusted': True, '_vts': 1720035544085, 'key': 'Enter', 'code': 'Enter', 'location': 0, 'ctrlKey': False, 'shiftKey': False, 'altKey': False, 'metaKey': False, 'repeat': False, 'isComposing': False, 'charCode': 13, 'keyCode': 13, 'DOM_KEY_LOCATION_STANDARD': 0, 'DOM_KEY_LOCATION_LEFT': 1, 'DOM_KEY_LOCATION_RIGHT': 2, 'DOM_KEY_LOCATION_NUMPAD': 3, 'detail': 0, 'sourceCapabilities': {}, 'which': 13, 'type': 'keypress', 'eventPhase': 2, 'bubbles': True, 'cancelable': True, 'defaultPrevented': False, 'composed': True, 'timeStamp': 2796.5999999940395, 'returnValue': True, 'cancelBubble': False, 'NONE': 0, 'CAPTURING_PHASE': 1, 'AT_TARGET': 2, 'BUBBLING_PHASE': 3}}
{'id': 4, 'client_id': 'a48113b2-a22d-427d-8bbf-601e4e0500fe', 'listener_id': '72ed4040-9518-4a02-b7f1-b13400fca0a5', 'args': {'isTrusted': True, '_vts': 1720035544085, 'key': 'Enter', 'code': 'Enter', 'location': 0, 'ctrlKey': False, 'shiftKey': False, 'altKey': False, 'metaKey': False, 'repeat': False, 'isComposing': False, 'charCode': 13, 'keyCode': 13, 'DOM_KEY_LOCATION_STANDARD': 0, 'DOM_KEY_LOCATION_LEFT': 1, 'DOM_KEY_LOCATION_RIGHT': 2, 'DOM_KEY_LOCATION_NUMPAD': 3, 'detail': 0, 'sourceCapabilities': {}, 'which': 13, 'type': 'keypress', 'eventPhase': 3, 'bubbles': True, 'cancelable': True, 'defaultPrevented': False, 'composed': True, 'timeStamp': 2796.5999999940395, 'returnValue': True, 'cancelBubble': False, 'NONE': 0, 'CAPTURING_PHASE': 1, 'AT_TARGET': 2, 'BUBBLING_PHASE': 3}}
{'id': 4, 'client_id': 'a48113b2-a22d-427d-8bbf-601e4e0500fe', 'listener_id': '650d15a8-6afe-43bf-bfc5-72aa7bdb6627', 'args': 'a<div><br></div>'}
```
Note that the previous code block is the result of me pressing the letter 'a' once, then enter. 

So how do we fix this? 
Ultimately, I have no idea. We need to change the way that event handlers are ordered, and I'm not sure of how it's currently being done.

**Comment by python-and-fiction** [comment]
It's interesting. `editor`'s init value is `123`.

First, input anything in editor and click `set value to 123(same)`,editor's display value(the text in editor) won't change.
but after reloaded, input anything  or not in editor and click `set value to qwe`,editor's display value  changes.
It seems that set_value to different value everytime is ok.
 
let's restart program.click `loopback to false`,input anything in editor and click `set value to 123(same)` once,
then input anything in editor and click `set value to 123(same)` ,editor's display value changes as expected but cursor jumps to start everytime.
if you click `loopback to none`,input anything in editor and click `set value to 123(same)` once,
then input anything in editor and click `set value to 123(same)` ,editor's display value won't change.
It seems that bug comes from events handler which changes value of editor ( mabyde in nicegui of quasar or vue? i'm not sure )

```python3
from nicegui import ui

strs = ['123','qwe']

editor = ui.editor(value=strs[0])

def set_loopback_false(editor):
    editor._props['loopback'] = False
def set_loopback_none(editor):
    editor._props['loopback'] = None
with ui.card():
    with ui.row():
        ui.button(f'set value to {strs[0]}(same)',on_click=lambda :editor.set_value(strs[0]))
        ui.button(f'set value to {strs[1]}',on_click=lambda :editor.set_value(strs[1]))
    with ui.row():
        ui.button(f'loopback to false ',on_click=lambda :set_loopback_false(editor))
        ui.button(f'loopback to none ',on_click=lambda :set_loopback_none(editor))
    with ui.row():
        ui.button('print editor props',on_click=lambda :ui.notify(editor._props))
        ui.button('print editor value',on_click=lambda :ui.notify(editor.value))

ui.run()
```
![image](https://github.com/zauberzeug/nicegui/assets/16256910/471eb033-519e-4eb5-ba54-9c58484292a5)

**Comment by falkoschindler** [comment]
As reported in #3337, there seems to be a similar problem with `ui.codemirror`:
```py
code = ui.codemirror(value='foo')
ui.label().bind_text_from(code, 'value')
ui.button('clear', on_click=lambda: code.set_value(''))

editor = ui.editor(value='foo')
ui.label().bind_text_from(editor, 'value')
ui.button('clear', on_click=lambda: editor.set_value(''))
```

**Comment by frankvp11** [comment]
Do you have any leads on what the problem might be? I'd like to help but I have no idea

**Comment by falkoschindler** [comment]
@frankvp11 I'm still investigating. Here are my current thoughts:

- Let's focus on `ui.editor`, because `ui.codemirror` does some more Vue magic on the client.
- Clearing the editor is just a special case for setting its value.
- I don't think it's a race condition, because we can decouple both events using a timer:

```py
editor = ui.editor(value='foo')
ui.label().bind_text_from(editor, 'value')
ui.button('set x', on_click=lambda: ui.timer(0.5, lambda: editor.set_value('x'), once=True))
```

**Comment by falkoschindler** [comment]
Another insight: Calling `set_value` with alternating values works.

```py
code = ui.codemirror(value='foo')
ui.label().bind_text_from(code, 'value')
ui.button('set x', on_click=lambda: ui.timer(0.5, lambda: code.set_value('x'), once=True))
ui.button('set y', on_click=lambda: ui.timer(0.5, lambda: code.set_value('y'), once=True))
```

I think in case of `ui.codemirror` the `value` prop doesn't change when typing, so the change detection doesn't trigger when the server sends the same "x" again:

https://github.com/zauberzeug/nicegui/blob/b76c2d44e2c46ed40337c947eba5304174354e07/nicegui/elements/codemirror.js#L16-L18

But `ui.editor` is implemented differently. So maybe the bug for `ui.editor` and `ui.codemirror` isn't the same after all.

**Comment by falkoschindler** [comment]
I found a solution for `ui.editor`, see PR #3346.
Since `ui.codemirror` needs to be addressed separately, I'll re-open #3337 again.
